ocaml/xapi/xapi_vm.ml

```diff
@@ -1019,6 +1026,48 @@ let set_auto_update_drivers ~__context ~self ~value=
 	assert_can_set_auto_update_drivers ~__context ~self ~value;
 	Db.VM.set_auto_update_drivers ~__context ~self ~value
 
+let xenprep_mutex = Mutex.create ()
+
 let xenprep_start ~__context ~self =
-	info "Xapi_vm.xenprep_start: VM=%s" (Db.VM.get_uuid ~__context ~self);
-	Db.VM.add_to_other_config ~__context ~self ~key:"xenprep_progress" ~value:"dummyvalue_should_be_cd_inserted"
+	let key = "xenprep_progress" in
+	let vm_uuid = Db.VM.get_uuid ~__context ~self in (* Just for log-msgs *)
+	info "Xapi_vm.xenprep_start: VM=%s" vm_uuid;
+	let started_already = Mutex.execute xenprep_mutex (fun () ->
+		if List.mem_assoc key (Db.VM.get_other_config ~__context ~self) then true
+		else (Db.VM.add_to_other_config ~__context ~self ~key ~value:"about_to_insert_iso";
+		false)
+	) in
+	if started_already then
+		info "Xapi_vm.xenprep_start: VM is in xenprep already: VM=%s" vm_uuid
+	else (
+		let cd_name = Xapi_globs.xenprep_iso_name_label in
+		let sr_name = Xapi_globs.miami_tools_sr_name in
+		let vdis = Db.VDI.get_refs_where ~__context ~expr:(
+			Db_filter_types.Eq (
+				Db_filter_types.Field "name_label",
+				Db_filter_types.Literal cd_name )
+		) in
+		let vdis = List.filter (fun vdi ->
+			let sr = Db.VDI.get_SR ~__context ~self:vdi in
+			("iso" = Db.SR.get_content_type ~__context ~self:sr)
+			&& (sr_name = Db.SR.get_name_label ~__context ~self:sr)
+			&& (Helpers.is_tools_sr ~__context ~sr)
+		) vdis in
+		if List.length vdis <> 1 then failwith
+			("xenprep_start failed: found "^(string_of_int (List.length vdis))^" ISOs with name-label="^cd_name^" in tools SR with name-label="^sr_name);
+		let vdi = List.hd vdis in
+		(* Find any empty CD-drive VBD for the VM. *)
+		(* We don't care if currently_attached is false: VBD.insert will take care of things. *)
+		let vbds = Db.VM.get_VBDs ~__context ~self in
+		let empty_cd_vbds = List.filter (fun vbd -> (Db.VBD.get_type ~__context ~self:vbd = `CD) && (Db.VBD.get_empty ~__context ~self:vbd)) vbds in
+		if [] = empty_cd_vbds then raise (Api_errors.Server_error(Api_errors.vm_no_empty_cd_vbd, [ Ref.string_of self ]));
+		let cd_vbd = List.hd empty_cd_vbds in
+		(* We found a suitable drive, so insert the CD. *)
+		Helpers.call_api_functions ~__context
```

Comment:
On second thoughts, I think I can replace this with just
`Xapi_vbd.insert ~__context ~vbd:cd_vbd ~vdi;`
instead of using `Helpers.call_api_functions`.